File: plugins/harness/skills/harness-install/assets/bun/runtime/rules/leaf-function-blank-lines.ts
#!/usr/bin/env bun
import type { FunctionLike, Node, SourceFile } from "typescript@6.0.3";
import {
	createSourceFile,
	forEachChild,
	isBlock,
	isFunctionDeclaration,
	isIdentifier,
	isMethodDeclaration,
	ScriptTarget,
	SyntaxKind,
} from "typescript@6.0.3";
import type {
	Finding,
	HarnessCheckRule,
	HarnessManifest,
	RuleContext,
} from "../harness-check-rule";

const CATEGORY = "leafFunctionBlankLines";

/**
 * Helper to check if a function has nested functions.
 */
const hasNestedFunctions = (funcNode: FunctionLike): boolean => {
	let found = false;
	const visit = (node: Node): void => {
		if (found) {
			return;
		}
		if (node !== funcNode) {
			switch (node.kind) {
				case SyntaxKind.FunctionDeclaration:
				case SyntaxKind.MethodDeclaration:
				case SyntaxKind.FunctionExpression:
				case SyntaxKind.ArrowFunction:
				case SyntaxKind.Constructor: {
					found = true;
					return;
				}
			}
		}
		forEachChild(node, visit);
	};
	forEachChild(funcNode, visit);
	return found;
};

export const leafFunctionBlankLinesRule = (
	ctx: RuleContext,
): HarnessCheckRule => ({
	category: "leafFunctionBlankLines",
	applies(_manifest: HarnessManifest): boolean {
		return true;
	},

	validate(_root: string, manifest: HarnessManifest): readonly Finding[] {
		const parameters = ctx.readJsonObject(
			ctx.readJsonObject(manifest[CATEGORY]).parameters,
		);
		const configuredMaxConsecutiveBlankLines = Number(
			parameters.maxConsecutiveBlankLines ?? 1,
		);
		const maxConsecutiveBlankLines = Number.isFinite(
			configuredMaxConsecutiveBlankLines,
		)
			? Math.max(0, Math.trunc(configuredMaxConsecutiveBlankLines))
			: 1;
		return ctx
			.stackSources(manifest, CATEGORY, "typescript")
			.flatMap((file) => findingsForFile(ctx, manifest, file, maxConsecutiveBlankLines));
	},
});

const findingsForFile = (
	ctx: RuleContext,
	manifest: HarnessManifest,
	file: string,
	maxConsecutiveBlankLines: number,
): readonly Finding[] => {
	const text = ctx.read(file);
	if (!text) {
		return [];
	}
	let sourceFile: SourceFile;
	try {
		sourceFile = createSourceFile(file, text, ScriptTarget.Latest, true);
	} catch {
		return [
			{
				severity: ctx.severityOf(manifest, CATEGORY),
				category: CATEGORY,
				message: `failed to parse TypeScript: ${file}`,
			},
		];
	}
	const findings: Finding[] = [];
	const visit = (node: Node): void => {
		switch (node.kind) {
			case SyntaxKind.FunctionDeclaration:
			case SyntaxKind.MethodDeclaration:
			case SyntaxKind.FunctionExpression:
			case SyntaxKind.ArrowFunction:
			case SyntaxKind.Constructor: {
				const funcLike = node as FunctionLike;
				if (funcLike.body && !hasNestedFunctions(funcLike)) {
					findings.push(
						...extractBlankLineFindings(
							ctx,
							manifest,
							file,
							text,
							sourceFile,
							funcLike,
							funcLike.body,
							maxConsecutiveBlankLines,
						),
					);
				}
				break;
			}
		}
		forEachChild(node, visit);
	};
	visit(sourceFile);
	return findings;
};

const extractBlankLineFindings = (
	ctx: RuleContext,
	manifest: HarnessManifest,
	file: string,
	text: string,
	sourceFile: SourceFile,
	funcNode: FunctionLike,
	body: FunctionLike["body"],
	maxConsecutiveBlankLines: number,
): readonly Finding[] => {
	const blankLineFindings: Finding[] = [];
	if (!isBlock(body) || body.statements.length === 0) {
		return blankLineFindings;
	}
	const funcName =
		(isFunctionDeclaration(funcNode) && funcNode.name?.text) ||
		(isMethodDeclaration(funcNode) &&
			funcNode.name &&
			isIdentifier(funcNode.name) &&
			funcNode.name.text) ||
		"<anonymous>";
	const startLine = sourceFile.getLineAndCharacterOfPosition(
		body.getStart(sourceFile, true),
	).line;
	const endLine = sourceFile.getLineAndCharacterOfPosition(body.getEnd()).line;
	let blankLines = 0;
	text
		.split(/\r?\n/)
		.slice(startLine, endLine + 1)
		.forEach((line, index) => {
			if (line.trim() === "") {
				blankLines += 1;
				if (blankLines > maxConsecutiveBlankLines) {
					blankLineFindings.push({
						severity: ctx.severityOf(manifest, CATEGORY),
						category: CATEGORY,
						message: `${file}:${startLine + index + 1}: leaf function \`${funcName}\` contains too many blank lines; remove or extract the section`,
					});
				}
			} else {
				blankLines = 0;
			}
		});
	return blankLineFindings;
};

File: plugins/harness/skills/harness-install/assets/uv/runtime/rules/leaf_function_blank_lines.py
#!/usr/bin/env -S uv run
# -*- coding: utf-8 -*-
"""Forbid multiple consecutive blank lines in leaf function rule."""

from collections.abc import Iterable

from pathlib import Path

import libcst as cst

from harness_check_rule import Finding, HarnessCheckRule

from .utils import has_nested_function, parse_python, relative, severity_for, stack_sources

class LeafFunctionBlankLinesRule(HarnessCheckRule):
    """Validate leafFunctionBlankLines check."""

    category = "leafFunctionBlankLines"

    def applies(self, manifest: dict) -> bool:
        """Check if this rule applies to the manifest."""
        section = manifest.get(self.category)
        if not isinstance(section, dict):
            return False
        return section.get("enabled", True) is not False

    def validate(self, root: Path, manifest: dict) -> Iterable[Finding]:
        """Validate leafFunctionBlankLines check."""
        severity = severity_for(manifest, self.category)
        category = self.category
        max_consecutive_blank_lines = self.max_consecutive_blank_lines(manifest)

        class BlankLineFinder(cst.CSTVisitor):
            METADATA_DEPENDENCIES = (cst.metadata.PositionProvider,)

            def __init__(self, rel_path: str) -> None:
                super().__init__()
                self.findings: list[Finding] = []
                self.rel_path = rel_path
            def visit_FunctionDef(self, node: cst.FunctionDef) -> bool:
                if has_nested_function(node):
                    return False
                if not isinstance(node.body, cst.IndentedBlock):
                    return False
                blank_lines = 0
                for stmt in node.body.body:
                    for line in stmt.leading_lines:
                        if line.comment is None:
                            blank_lines += 1
                            if max_consecutive_blank_lines < blank_lines:
                                pos = self.get_metadata(cst.metadata.PositionProvider, line)
                                self.findings.append(Finding(
                                    severity,
                                    category,
                                    f"{self.rel_path}:{pos.start.line}: leaf function `{node.name.value}` "
                                    "contains too many blank lines; remove or extract the section",
                                ))
                        else:
                            blank_lines = 0
                    blank_lines = 0
                return True
        findings = []
        for path in stack_sources(root, manifest, self.category):
            tree, error = parse_python(path)
            if error is not None:
                findings.append(Finding(
                    severity,
                    self.category,
                    f"{relative(path)}: syntax error: {error}",
                ))
                continue
            wrapper = cst.MetadataWrapper(tree)
            visitor = BlankLineFinder(relative(path))
            wrapper.visit(visitor)
            findings.extend(visitor.findings)
        return findings

    def max_consecutive_blank_lines(self, manifest: dict) -> int:
        """Return the maximum allowed consecutive blank lines."""
        section = manifest.get(self.category)
        if not isinstance(section, dict):
            return 1
        parameters = section.get("parameters")
        if not isinstance(parameters, dict):
            return 1
        value = parameters.get("maxConsecutiveBlankLines")
        return max(0, value) if isinstance(value, int) else 1


RULE: HarnessCheckRule = LeafFunctionBlankLinesRule()
